File: src/timerapp_ag/webdav_sync.py
"""Синхронизация data.json с WebDAV: merge, versioning, pull-before-push."""
from __future__ import annotations

import json
import logging
import tempfile
import time
from dataclasses import dataclass
from pathlib import Path

from .domain.sync_normalize import normalize_running_tasks
from .storage import AppState, Storage, merge_data_files
from .webdav_client import WebDavClient, WebDavError
from .webdav_config import (
    WebDavConfig,
    load_webdav_config,
    mark_webdav_sync_error,
    mark_webdav_sync_ok,
    save_webdav_config,
    save_webdav_pending_notice,
)
from .webdav_meta import RemoteSyncMeta, content_hash, meta_to_bytes, new_meta, parse_meta_bytes

logger = logging.getLogger(__name__)


@dataclass
class SyncOutcome:
    state: AppState | None = None
    error: str = ""
    conflict_detected: bool = False
    notice: str = ""


def _read_remote_meta(client: WebDavClient, config: WebDavConfig) -> RemoteSyncMeta | None:
    meta_url = config.meta_remote_url()
    if not client.exists(meta_url):
        return None
    try:
        payload = client.download(meta_url)
    except WebDavError as exc:
        if exc.status_code == 404:
            return None
        raise
    return parse_meta_bytes(payload)


_META_UPLOAD_ATTEMPTS = 3
_FULL_UPLOAD_CYCLES = 2


def _remote_payload_hash(payload: bytes, meta: RemoteSyncMeta | None) -> str:
    """SHA-256 содержимого; при рассинхроне meta и файла — доверяем файлу."""
    file_hash = content_hash(payload)
    if meta is None:
        return file_hash
    if meta.content_hash != file_hash:
        logger.warning(
            "WebDAV sync-meta hash mismatch (meta=%s, file=%s); using file hash",
            meta.content_hash[:12],
            file_hash[:12],
        )
        return file_hash
    return meta.content_hash


def _upload_meta_with_retries(
    client: WebDavClient,
    meta_url: str,
    meta_bytes: bytes,
) -> None:
    last_error: WebDavError | None = None
    for attempt in range(1, _META_UPLOAD_ATTEMPTS + 1):
        try:
            client.upload(meta_url, meta_bytes)
            return
        except WebDavError as exc:
            last_error = exc
            logger.warning(
                "WebDAV sync-meta upload attempt %s/%s failed: %s",
                attempt,
                _META_UPLOAD_ATTEMPTS,
                exc,
            )
            if attempt < _META_UPLOAD_ATTEMPTS:
                time.sleep(0.5 * attempt)
    assert last_error is not None
    raise last_error


def _upload_payload(client: WebDavClient, config: WebDavConfig, payload: bytes) -> RemoteSyncMeta:
    data_url = config.remote_url()
    meta_url = config.meta_remote_url()
    device_id = config.ensure_device_id()
    meta = new_meta(payload, device_id)
    meta_bytes = meta_to_bytes(meta)
    last_error: WebDavError | None = None
    for cycle in range(1, _FULL_UPLOAD_CYCLES + 1):
        try:
            client.upload(data_url, payload)
            _upload_meta_with_retries(client, meta_url, meta_bytes)
            return meta
        except WebDavError as exc:
            last_error = exc
            logger.warning(
                "WebDAV upload cycle %s/%s failed: %s",
                cycle,
                _FULL_UPLOAD_CYCLES,
                exc,
            )
            if cycle < _FULL_UPLOAD_CYCLES:
                time.sleep(0.5 * cycle)
    assert last_error is not None
    raise WebDavError(
        f"Не удалось загрузить data.json и sync-meta после {_FULL_UPLOAD_CYCLES} циклов: {last_error}"
    ) from last_error


def _merge_local_with_remote(storage: Storage, remote_payload: bytes) -> AppState:
    try:
        json.loads(remote_payload.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise WebDavError("Удалённый файл не является корректным JSON") from exc

    with tempfile.NamedTemporaryFile("wb", suffix=".json", delete=False) as handle:
        handle.write(remote_payload)
        remote_path = Path(handle.name)
    try:
        candidates = [storage.path, remote_path]
        return merge_data_files(candidates)
    finally:
        remote_path.unlink(missing_ok=True)


def _finalize_merged_state(storage: Storage, merged: AppState, *, reason: str) -> AppState:
    normalize_running_tasks(merged)
    storage.save(merged, update_rolling_backup=False)
    storage.create_backup(reason)
    return merged


def _remote_changed_since_sync(
    config: WebDavConfig,
    remote_hash: str,
    local_hash: str,
) -> bool:
    """Удалённый файл изменился с прошлой синхронизации или отличается от локального."""
    if config.last_remote_content_hash:
        return remote_hash != config.last_remote_content_hash
    return bool(local_hash) and remote_hash != local_hash


def pull_and_merge(
    storage: Storage,
    config: WebDavConfig | None = None,
    *,
    require_enabled: bool = True,
) -> SyncOutcome:
    """Скачать удалённую базу и объединить с локальной."""
    config = config or load_webdav_config()
    if require_enabled and not config.enabled:
        raise WebDavError("Синхронизация WebDAV отключена")

    client = WebDavClient(config)
    conflict_detected = False
    candidates = [storage.path]

    if client.exists():
        remote_payload = client.download()
        remote_meta = _read_remote_meta(client, config)
        remote_hash = _remote_payload_hash(remote_payload, remote_meta)
        local_hash = content_hash(storage.path.read_bytes()) if storage.path.is_file() else ""
        if _remote_changed_since_sync(config, remote_hash, local_hash):
            conflict_detected = True
            logger.warning(
                "WebDAV conflict: remote hash changed (%s -> %s)",
                (config.last_remote_content_hash or local_hash)[:12],
                remote_hash[:12],
            )
        merged = _merge_local_with_remote(storage, remote_payload)
    else:
        merged = merge_data_files(candidates) if storage.path.exists() else AppState()

    merged = _finalize_merged_state(storage, merged, reason="webdav-pull")
    remote_hash = content_hash(storage.path.read_bytes())
    mark_webdav_sync_ok(config, remote_hash=remote_hash, had_conflict=conflict_detected)

    notice = ""
    if conflict_detected:
        notice = "Обнаружен конфликт версий: данные объединены с сервера."

    return SyncOutcome(state=merged, conflict_detected=conflict_detected, notice=notice)


def push_local(
    storage: Storage,
    config: WebDavConfig | None = None,
    *,
    require_enabled: bool = True,
) -> SyncOutcome:
    """Pull-before-push: merge с сервером, затем upload data.json + sync-meta."""
    config = config or load_webdav_config()
    if require_enabled and not config.enabled:
        raise WebDavError("Синхронизация WebDAV отключена")
    if not storage.path.is_file():
        raise WebDavError("Локальный файл данных не найден")

    client = WebDavClient(config)
    conflict_detected = False

    if client.exists():
        remote_payload = client.download()
        remote_meta = _read_remote_meta(client, config)
        remote_hash = _remote_payload_hash(remote_payload, remote_meta)
        local_hash = content_hash(storage.path.read_bytes())
        if _remote_changed_since_sync(config, remote_hash, local_hash):
            conflict_detected = True
            logger.warning("WebDAV push: remote changed since last sync, merging before upload")
        merged = _merge_local_with_remote(storage, remote_payload)
        _finalize_merged_state(storage, merged, reason="webdav-push-merge")
    else:
        merged = AppState.from_dict(json.loads(storage.path.read_text(encoding="utf-8")))
        normalize_running_tasks(merged)
        storage.save(merged, update_rolling_backup=False)

    payload = storage.path.read_bytes()
    meta = _upload_payload(client, config, payload)
    mark_webdav_sync_ok(config, remote_hash=meta.content_hash, had_conflict=conflict_detected)

    notice = ""
    if conflict_detected:
        notice = "Перед загрузкой выполнено слияние с более новой версией на сервере."

    return SyncOutcome(state=merged, conflict_detected=conflict_detected, notice=notice)


def push_local_upload_only(
    storage: Storage,
    config: WebDavConfig | None = None,
    *,
    require_enabled: bool = True,
) -> SyncOutcome:
    """Отправить локальный data.json без слияния с сервером (опция при выходе)."""
    config = config or load_webdav_config()
    if require_enabled and not config.enabled:
        raise WebDavError("Синхронизация WebDAV отключена")
    if not storage.path.is_file():
        raise WebDavError("Локальный файл данных не найден")

    client = WebDavClient(config)
    conflict_detected = False
    local_payload = storage.path.read_bytes()
    local_hash = content_hash(local_payload)

    if client.exists():
        remote_payload = client.download()
        remote_meta = _read_remote_meta(client, config)
        remote_hash = _remote_payload_hash(remote_payload, remote_meta)
        if _remote_changed_since_sync(config, remote_hash, local_hash):
            conflict_detected = True
            logger.warning(
                "WebDAV upload-only: remote changed since last sync (%s -> %s)",
                (config.last_remote_content_hash or local_hash)[:12],
                remote_hash[:12],
            )

    payload = local_payload
    meta = _upload_payload(client, config, payload)
    mark_webdav_sync_ok(config, remote_hash=meta.content_hash, had_conflict=conflict_detected)

    notice = ""
    if conflict_detected:
        notice = (
            "При выходе на сервер отправлена локальная копия без слияния; "
            "в облаке была более новая версия."
        )

    return SyncOutcome(conflict_detected=conflict_detected, notice=notice)


def _persist_shutdown_notice(outcome: SyncOutcome) -> None:
    if outcome.notice and outcome.conflict_detected:
        save_webdav_pending_notice(outcome.notice)


def sync_webdav_on_shutdown(storage: Storage) -> SyncOutcome:
    config = load_webdav_config()
    if not config.enabled or not config.sync_on_shutdown:
        return SyncOutcome()
    try:
        if config.shutdown_upload_only:
            outcome = push_local_upload_only(storage, config, require_enabled=True)
        else:
            outcome = push_local(storage, config, require_enabled=True)
        _persist_shutdown_notice(outcome)
        return outcome
    except WebDavError as exc:
        logger.error("WebDAV shutdown sync failed: %s", exc)
        mark_webdav_sync_error(config, str(exc))
        return SyncOutcome(error=str(exc))


def test_webdav_connection(config: WebDavConfig) -> str:
    client = WebDavClient(config)
    if client.exists():
        return client.test_connection() + " (удалённый файл найден)"
    client._ensure_collection(config.remote_url())
    return "Подключение успешно (удалённый файл будет создан при первой синхронизации)"


def sync_webdav_on_startup(storage: Storage) -> SyncOutcome:
    config = load_webdav_config()
    if not config.enabled or not config.sync_on_startup:
        return SyncOutcome()
    try:
        return pull_and_merge(storage, config, require_enabled=True)
    except WebDavError as exc:
        logger.error("WebDAV startup sync failed: %s", exc)
        mark_webdav_sync_error(config, str(exc))
        return SyncOutcome(error=str(exc))


def save_webdav_settings(config: WebDavConfig) -> None:
    save_webdav_config(config)

File: android/app/src/main/java/com/timerapp/linkb24/webdav/WebDavSync.kt
package com.timerapp.linkb24.webdav

import com.timerapp.linkb24.data.AppDataDto
import com.timerapp.linkb24.data.TaskRepository
import com.timerapp.linkb24.data.WebDavConfig
import com.timerapp.linkb24.data.WebDavConfigRepository
import com.timerapp.linkb24.data.mergeDataFiles
import com.timerapp.linkb24.data.normalizeRunningTasks

data class SyncOutcome(
    val data: AppDataDto? = null,
    val error: String = "",
    val conflictDetected: Boolean = false,
    val notice: String = "",
)

data class RemoteCheckOutcome(
    val remoteChanged: Boolean = false,
    val remoteHash: String = "",
    val error: String = "",
)

class WebDavSync(
    private val taskRepository: TaskRepository,
    private val configRepository: WebDavConfigRepository,
) {
    fun syncOnStartup(): SyncOutcome {
        val config = configRepository.load()
        if (!config.enabled || !config.syncOnStartup) {
            return SyncOutcome()
        }
        return runCatching {
            pullAndMerge(config, requireEnabled = true)
        }.getOrElse { error ->
            val message = syncErrorMessage(error)
            configRepository.markSyncError(config, message)
            SyncOutcome(error = message)
        }
    }

    fun syncOnShutdown(): SyncOutcome {
        val config = configRepository.load()
        if (!config.enabled || !config.syncOnShutdown) {
            return SyncOutcome()
        }
        return runCatching {
            val outcome = if (config.shutdownUploadOnly) {
                pushLocalUploadOnly(config, requireEnabled = true)
            } else {
                pushLocal(config, requireEnabled = true)
            }
            if (outcome.notice.isNotBlank() && outcome.conflictDetected) {
                configRepository.savePendingNotice(outcome.notice)
            }
            outcome
        }.getOrElse { error ->
            val message = syncErrorMessage(error)
            configRepository.markSyncError(config, message)
            SyncOutcome(error = message)
        }
    }

    fun checkRemoteOnPeriodic(): RemoteCheckOutcome {
        val config = configRepository.load()
        if (!config.periodicSyncEnabled()) {
            return RemoteCheckOutcome()
        }
        return checkRemoteChanges(config, requireEnabled = true)
    }

    fun checkRemoteChanges(
        config: WebDavConfig = configRepository.load(),
        requireEnabled: Boolean = true,
    ): RemoteCheckOutcome {
        if (requireEnabled && !config.enabled) {
            return RemoteCheckOutcome()
        }
        if (!config.isConfigured()) {
            return RemoteCheckOutcome(error = "WebDAV не настроен: укажите URL и имя пользователя")
        }
        return runCatching {
            val client = WebDavClient(config.withDeviceId())
            if (!client.exists()) {
                return RemoteCheckOutcome()
            }
            val remotePayload = client.download()
            val remoteMeta = readRemoteMeta(client, config)
            val remoteHash = remotePayloadHash(remotePayload, remoteMeta)
            val dataFile = taskRepository.dataFile
            val localHash = if (dataFile.isFile) contentHash(dataFile.readBytes()) else ""
            val changed = remoteChangedSinceSync(config, remoteHash, localHash)
            RemoteCheckOutcome(remoteChanged = changed, remoteHash = remoteHash)
        }.getOrElse { error ->
            RemoteCheckOutcome(error = syncErrorMessage(error))
        }
    }

    fun syncNow(
        config: WebDavConfig = configRepository.load(),
        requireEnabled: Boolean = false,
    ): SyncOutcome {
        if (requireEnabled && !config.enabled) {
            return SyncOutcome(error = "Синхронизация WebDAV отключена")
        }
        if (!config.isConfigured()) {
            return SyncOutcome(error = "WebDAV не настроен: укажите URL и имя пользователя")
        }
        return runCatching {
            val pullOutcome = pullAndMerge(config, requireEnabled = false)
            val pushOutcome = pushMerged(config, requireEnabled = false)
            val notices = listOf(pullOutcome.notice, pushOutcome.notice).filter { it.isNotBlank() }
            SyncOutcome(
                data = pushOutcome.data ?: pullOutcome.data,
                conflictDetected = pullOutcome.conflictDetected || pushOutcome.conflictDetected,
                notice = notices.joinToString(" ").ifBlank { "Синхронизация завершена" },
            )
        }.getOrElse { error ->
            val message = syncErrorMessage(error)
            configRepository.markSyncError(config, message)
            SyncOutcome(error = message)
        }
    }

    fun pullAndMerge(
        config: WebDavConfig = configRepository.load(),
        requireEnabled: Boolean = true,
    ): SyncOutcome {
        if (requireEnabled && !config.enabled) {
            throw WebDavException("Синхронизация WebDAV отключена")
        }
        val client = WebDavClient(config.withDeviceId())
        var conflictDetected = false
        val dataFile = taskRepository.dataFile
        val remoteFound = client.exists()

        val merged = if (remoteFound) {
            val remotePayload = client.download()
            val remoteMeta = readRemoteMeta(client, config)
            val remoteHash = remotePayloadHash(remotePayload, remoteMeta)
            val localHash = if (dataFile.isFile) contentHash(dataFile.readBytes()) else ""
            if (remoteChangedSinceSync(config, remoteHash, localHash)) {
                conflictDetected = true
            }
            mergeDataFiles(dataFile, remotePayload)
        } else if (dataFile.isFile) {
            taskRepository.load()
        } else {
            AppDataDto()
        }

        val finalized = finalizeMergedState(merged)
        val remoteHash = contentHash(dataFile.readBytes())
        configRepository.markSyncOk(config.withDeviceId(), remoteHash, conflictDetected)

        val notice = when {
            conflictDetected -> "Обнаружен конфликт версий: данные объединены с сервера."
            !remoteFound && merged.tasks.isEmpty() ->
                "Файл ${config.remotePath} на сервере не найден. " +
                    "Сначала загрузите data.json с компьютера или нажмите «Загрузить сейчас»."
            !remoteFound ->
                "Файл ${config.remotePath} на сервере не найден — использована локальная копия."
            else -> ""
        }
        return SyncOutcome(
            data = finalized,
            conflictDetected = conflictDetected,
            notice = notice,
        )
    }

    fun pushLocal(
        config: WebDavConfig = configRepository.load(),
        requireEnabled: Boolean = true,
    ): SyncOutcome {
        if (requireEnabled && !config.enabled) {
            throw WebDavException("Синхронизация WebDAV отключена")
        }
        val dataFile = taskRepository.dataFile
        if (!dataFile.isFile) {
            throw WebDavException("Локальный файл данных не найден")
        }

        val client = WebDavClient(config.withDeviceId())
        var conflictDetected = false
        var merged = taskRepository.load()

        if (client.exists()) {
            val remotePayload = client.download()
            val remoteMeta = readRemoteMeta(client, config)
            val remoteHash = remotePayloadHash(remotePayload, remoteMeta)
            val localHash = contentHash(dataFile.readBytes())
            if (remoteChangedSinceSync(config, remoteHash, localHash)) {
                conflictDetected = true
            }
            merged = mergeDataFiles(dataFile, remotePayload)
            finalizeMergedState(merged)
        } else {
            merged = normalizeRunningTasks(merged)
            taskRepository.save(merged)
        }

        val payload = dataFile.readBytes()
        val meta = uploadPayload(client, config.withDeviceId(), payload)
        configRepository.markSyncOk(config.withDeviceId(), meta.contentHash, conflictDetected)

        val notice = if (conflictDetected) {
            "Перед загрузкой выполнено слияние с более новой версией на сервере."
        } else {
            ""
        }
        return SyncOutcome(data = merged, conflictDetected = conflictDetected, notice = notice)
    }

    fun pushMerged(
        config: WebDavConfig = configRepository.load(),
        requireEnabled: Boolean = true,
    ): SyncOutcome {
        if (requireEnabled && !config.enabled) {
            throw WebDavException("Синхронизация WebDAV отключена")
        }
        val dataFile = taskRepository.dataFile
        if (!dataFile.isFile) {
            throw WebDavException("Локальный файл данных не найден")
        }

        val merged = normalizeRunningTasks(taskRepository.load())
        taskRepository.save(merged)
        val client = WebDavClient(config.withDeviceId())
        val payload = dataFile.readBytes()
        val meta = uploadPayload(client, config.withDeviceId(), payload)
        configRepository.markSyncOk(config.withDeviceId(), meta.contentHash, false)
        return SyncOutcome(data = merged)
    }

    fun pushLocalUploadOnly(
        config: WebDavConfig = configRepository.load(),
        requireEnabled: Boolean = true,
    ): SyncOutcome {
        if (requireEnabled && !config.enabled) {
            throw WebDavException("Синхронизация WebDAV отключена")
        }
        val dataFile = taskRepository.dataFile
        if (!dataFile.isFile) {
            throw WebDavException("Локальный файл данных не найден")
        }

        val client = WebDavClient(config.withDeviceId())
        var conflictDetected = false
        val localPayload = dataFile.readBytes()
        val localHash = contentHash(localPayload)

        if (client.exists()) {
            val remotePayload = client.download()
            val remoteMeta = readRemoteMeta(client, config)
            val remoteHash = remotePayloadHash(remotePayload, remoteMeta)
            if (remoteChangedSinceSync(config, remoteHash, localHash)) {
                conflictDetected = true
            }
        }

        val meta = uploadPayload(client, config.withDeviceId(), localPayload)
        configRepository.markSyncOk(config.withDeviceId(), meta.contentHash, conflictDetected)

        val notice = if (conflictDetected) {
            "При выходе на сервер отправлена локальная копия без слияния; " +
                "в облаке была более новая версия."
        } else {
            ""
        }
        return SyncOutcome(conflictDetected = conflictDetected, notice = notice)
    }

    private fun finalizeMergedState(merged: AppDataDto): AppDataDto {
        val normalized = normalizeRunningTasks(merged)
        taskRepository.save(normalized)
        return normalized
    }

    private fun readRemoteMeta(client: WebDavClient, config: WebDavConfig): RemoteSyncMeta? {
        val metaUrl = config.metaRemoteUrl()
        if (!client.exists(metaUrl)) {
            return null
        }
        return runCatching {
            parseMetaBytes(client.download(metaUrl))
        }.getOrNull()
    }

    private fun remoteChangedSinceSync(
        config: WebDavConfig,
        remoteHash: String,
        localHash: String,
    ): Boolean {
        if (config.lastRemoteContentHash.isNotBlank()) {
            return remoteHash != config.lastRemoteContentHash
        }
        return localHash.isNotEmpty() && remoteHash != localHash
    }

    private fun uploadPayload(
        client: WebDavClient,
        config: WebDavConfig,
        payload: ByteArray,
    ): RemoteSyncMeta {
        val dataUrl = config.remoteUrl()
        val metaUrl = config.metaRemoteUrl()
        val deviceId = config.withDeviceId().deviceId
        val meta = newMeta(payload, deviceId)
        val metaBytes = metaToBytes(meta)
        var lastError: WebDavException? = null
        repeat(FULL_UPLOAD_CYCLES) { cycle ->
            try {
                client.upload(dataUrl, payload)
                uploadMetaWithRetries(client, metaUrl, metaBytes)
                return meta
            } catch (error: WebDavException) {
                lastError = error
                if (cycle < FULL_UPLOAD_CYCLES - 1) {
                    Thread.sleep(500L * (cycle + 1))
                }
            }
        }
        throw WebDavException(
            "Не удалось загрузить data.json и sync-meta после $FULL_UPLOAD_CYCLES циклов: $lastError",
        )
    }

    private fun uploadMetaWithRetries(
        client: WebDavClient,
        metaUrl: String,
        metaBytes: ByteArray,
    ) {
        var lastError: WebDavException? = null
        repeat(META_UPLOAD_ATTEMPTS) { attempt ->
            try {
                client.upload(metaUrl, metaBytes)
                return
            } catch (error: WebDavException) {
                lastError = error
                if (attempt < META_UPLOAD_ATTEMPTS - 1) {
                    Thread.sleep(500L * (attempt + 1))
                }
            }
        }
        throw lastError ?: WebDavException("Не удалось загрузить sync-meta")
    }

    private fun syncErrorMessage(error: Throwable): String {
        return when (error) {
            is WebDavException -> error.message ?: "Ошибка WebDAV"
            is IllegalArgumentException -> error.message ?: "Ошибка данных"
            else -> error.message ?: error.javaClass.simpleName
        }
    }

    companion object {
        private const val META_UPLOAD_ATTEMPTS = 3
        private const val FULL_UPLOAD_CYCLES = 2
    }
}
